File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        m, n = len(grid), len(grid[0])

        def erase(i, j):
            if i < 0 or j < 0 or i >= m or j >= n or grid[i][j] != "1":
                return
            grid[i][j] = "0"
            erase(i + 1, j)
            erase(i - 1, j)
            erase(i, j + 1)
            erase(i, j - 1)

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    count += 1
                    erase(i, j)
        return count

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = m ? grid[0].size() : 0;
        int count = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    erase(grid, i, j);
                }
            }
        }
        return count;
    }

    void erase(vector<vector<char>>& grid, int i, int j) {
        int m = grid.size();
        int n = grid[0].size();

        if (i < 0 || j < 0 || i == m || j == n || grid[i][j] == '0') {
            return;
        }
        grid[i][j] = '0';
        erase(grid, i - 1, j);
        erase(grid, i + 1, j);
        erase(grid, i, j - 1);
        erase(grid, i, j + 1);
    }
};